[Scala] add support for annotations

gwenzek opened the pull request:
As mentioned in #588, I propose to add support for annotations.
I based myself on http://www.scala-lang.org/files/archive/spec/2.11/11-annotations.html .

I used the scope `storage.modifier.annotation.scala` on annotations.
This provide coloring in most themes due to `storage.modifier`and is easily customizable.

``` scala
@deprecated("Use D", "1.0") class C { ... }
// ^^ storage.modifier.annotation
//            ^^ string

@transient @volatile var m: Int
// ^^ storage.modifier.annotation

String @local
//     ^^ storage.modifier.annotation

(e: @unchecked) match { ... }
//  ^^ storage.modifier.annotation
//              ^^^^^ keyword.control.flow.scala

// more complex:
@scala.beans.BeanProperty
//  ^^^^^^^^^^^^^^^^^^^^^ storage.modifier.annotation

(e: Int @unchecked) match { ... }
//  ^^ storage.type.primitive
//      ^^ storage.modifier.annotation
//                  ^^^^^ keyword.control.flow.scala

trait Function0[@specialized(Unit, Int, Double) T] {
//               ^^ storage.modifier.annotation
//                           ^^ storage.type.primitive
//                                              ^ support.class
  def apply: T
}

```

**Note**
I match annotations with `{{plainid}}` regex, but we might want to support something more complex in the future.

I handle `@specialized` separately because it's the only annotation I know that uses type parameters between `(`.

djspiewak commented on Scala/Scala.sublime-syntax:
```diff
@@ -813,6 +857,7 @@ contexts:
       scope: support.type.scala
     - include: base-type-expression
   single-type-expression:
+    - include: annotation
```
This isn't going to work quite correctly, and also I think it's probably wrong.  For example:

```scala
x: @specialized Foo
```

Does that parse?  I actually don't know, but I don't think it does.  

Anyway, the reason it won't work correctly is because the space following `@specialized` will break out of the `single-type-expression` context!  If the above *does* parse, then you need to have a bit more machinery that parses the annotation, consumes the whitespace and then `set`s back into `single-type-expression`.  It isn't valid to ascribe *just* an annotation as a type, so that probably works.

gwenzek replied:
Thanks for the review.
I just tried with Scala 2.10 and 2.11 and `(e: Int @unchecked)` compiles but `(e: @unchecked Int)` doesn't ( "error: ')' expected but identifier found." )
The `annotation` context doesn't popup after the annotation and we still are in the `single-type-expression` so it works with  `(e: Int @unchecked)`.
With `(e: @unchecked Foo)` the `Foo` appears as a `support.constant` when the user would expect a type, which hints the compile error.

djspiewak replied:
I think it would be better to either highlight it as `invalid` or simply remove the highlighting for annotations inside of single-type-expression.

djspiewak commented on Scala/Scala.sublime-syntax:
```diff
@@ -130,6 +131,48 @@ contexts:
       push: single-type-expression
 
+  annotation:
+    - match: '@specialized'
```
This really shouldn't be special-cased.  First off, there's nothing particularly magical about `@specialized`, and other annotations *can* take type parameters in this manner.  Second, it is extremely common in practice to import-rename `@specialized` to `@sp` or `@spec`, for concision.  Your special-casing doesn't catch this.

I would rather simply support parenthetical parameters on annotations.  You'll have to do a bit of lookahead for it, I think, but that's not too bad.

gwenzek replied:
>  other annotations *can* take type parameters in this manner

I'm not really familiar with this kind of annotations. I mainly use annotations with string/int parameters. I haven't found documentation on annotations having types inside "()".
If you have pointers or code sample, I'll be happy to have a look at them.

My plan is to support  `@test[A, B]`, `@test(1, "foo")` and `@test(a, b)`
But I'm not sure how to handles `@test(Int, Foo)` because it's not really different from the previous.

djspiewak commented:
I agree that this fits well as a subscope of `storage.modifier`.  That's semantically closest to what it is, and fits well with how most color schemes use that scope.

djspiewak commented:
Here's a link to the way in which you can get types (or really anything) in annotation parens: http://docs.scala-lang.org/overviews/macros/annotations.html

djspiewak commented:
I think what you can probably do to handle annotation parentheticals:

```yaml
  - match: '@{{id}}'
    push: try-ann-params

try-ann-params:
  - match: '\('
    push:
      - match: '\)'
        pop: true
      - include: primitives
      - include: delimited-type-expression
  - match: '(?=\S)'
    pop: true
```

You'll need to extract numeric, boolean and string literals into a new context, which shouldn't take much refactoring.  Also, you'll want to match `classOf`, since that will result in a class literal rather than a type.

gwenzek commented:
> You'll need to extract numeric, boolean and string literals into a new context, `literals`

Isn't the context `constants` already doing that ?

gwenzek commented:
So I removed the special handling of `@specialized`, and "annotation" is not included in the "single-type-expression" context anymore.
The supported annotation parameters are any of the "constants" context `new ...` expressions and strings.

This allow to highlight properly the following edge cases:

```scala
@foo(new AnyRef { })
@foo(List(1, "foo"))

(e: Int @unchecked)
(e: @unchecked)

@specialized(Unit, Int, Double)
```

djspiewak commented:
> Isn't the context constants already doing that ?

Sort of.  It doesn't handle strings (which you want), and it *does* handle upper-case constants (which you don't want).  So I think we need a bit of reorganization there.

djspiewak commented:
> `@foo(new AnyRef { })`
> `@foo(List(1, "foo"))`

I don't believe that either of these are actually legal outside of macro annotations.  Do you feel it's worth highlighting those cases?

djspiewak commented on Scala/Scala.sublime-syntax:
```diff
@@ -867,6 +866,9 @@ contexts:
       scope: support.type.scala
       set: single-type-expression-tail
     - include: base-type-expression
+    - match: '(?=@)'
```
Why not just do this in `single-type-expression-tail`?  This is how we handle the following case:

```scala
e: Foo with Bar
```

djspiewak commented:
@gwenzek I gave it some thought (well, a lot of thought), and I think you should just allow arbitrary expressions within annotation parentheses.  The problem we have here is that macro annotations allow people to do *anything* and we can't really detect that.  This is a syntax highlighting mode, so we have to be really aggressive about what we highlight.

The only downside to allowing arbitrary expressions within annotation parentheticals is it means that the types in `@specialized(Unit, Int, Double)` will scope as constants rather than types.  I'm not sure that's particularly terrible (e.g. in Monokai, the only difference is the former is italic while the latter is not), but it is a thing.  Ultimately, we can't really tell what people *really* want there, and `@specialized` is specifically taking constants and remapping them as types through AST rewriting, so it's a bit unfair to expect Sublime to sort that all out.

So tl;dr: I would literally treat it as a `try-dispatch`.

gwenzek commented:
I agree with you, that's why I finally settled for matching 'strings', 'constants' and 'initialization'.
If you think it's better to push a 'main' context I can do it, but is it really allowed to put *anything* like a class definition in an annotation ?

djspiewak commented:
I don't think you can put a class definition in an annotation, but you can definitely put arbitrary expressions (e.g. ascriptions, lambdas, etc).

gwenzek commented:
Ok, I allowed anything as annotation parameter using `main` context, as you did in `try-dispatch`.

I think the PR is good to be merged, but maybe I could collapse the commits since I ended with 6 different commits for what is finally a minor code modification.

djspiewak commented:
@gwenzek Oh, btw, should we remove the special-casing of `@volatile` with this PR?

gwenzek commented:
@wbond The PR is in good shape now and has been validated by @djspiewak .
If you're okay, I'll collapse the commits into one and you'll be able to merge it !

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -1039,6 +1074,9 @@ contexts:
       set: single-type-expression-tail
     - include: base-type-expression
+    - match: '(?=@)'
+      comment: the '@' indicates the beginning of an annotation
```
With YAML defs we are moving away from `comment` keys and just writing a YAML comment.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
       push: single-type-expression
 
+  annotation:
+    - match: '@{{plainid}}\.'
```
The `@` symbol should probably have the scope `punctuation.definition.annotation.scala`, like how variables do in PHP.

djspiewak replied:
I'd like it if we could give them this scope, but *overlapping* with the `storage.modifier.annotation` scope, so that the `@` doesn't end up "randomly not colored" in most color schemes.

wbond replied:
Yes, that is correct, the `@` should be in addition to the `storage.modifier.annotation.scala` scope.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+        - match: '{{plainid}}'
+          scope: storage.modifier.annotation.scala
+          set: annotation-parameters
+    - match: '@{{plainid}}'
```
Add the `punctuation` scope to `@` here also.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+
+  annotation-parameters:
+    - match: \(
+      scope: punctuation.definition.arguments.annotation.begin.scala
```
This and related scopes should start with `punctuation.section.`. We are using `punctuation.definition` for more single-token/atomic type things and `punctuation.section` for larger areas, like braces, parens, etc.

This is an ongoing change, so there will be plenty of examples that still use `punctuation.definition` for this type of construct.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+  annotation:
+    - match: '@{{plainid}}\.'
+      scope: storage.modifier.annotation.scala
+      push:
```
Let's add a `meta.` scope for the annotation.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+      scope: storage.modifier.annotation.scala
+      push: annotation-parameters
+
+  annotation-parameters:
```
The same `meta.` scope should be added here as in the match in the `annotation` context.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+        - match: ','
+          scope: punctuation.separator.arguments.annotation.scala
+        - include: main
+        - match: \)
```
This match should probably be moved up above `include: main` in case a match for `\)` was added there in the future.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,39 @@ contexts:
+          scope: punctuation.definition.arguments.annotation.end.scala
+          pop: true
+        - include: delimited-type-expression
+    - match: (?!\(|\[)
```
Is whitespace allowed between `@identifier` and `(`? If so, this should allow `\s+` also.

wbond commented on Scala/syntax_test_scala.scala:
```diff
@@ -1189,3 +1189,38 @@ def <(a: Int) = 42
    </foo>
+
+// annotation examples from: http://www.scala-lang.org/files/archive/spec/2.11/11-annotations.html
+@deprecated("Use D", "1.0") class C { ... }
```
Add assertions to test the `@`, `,` and the `meta.` scope.

wbond replied:
This still needs to be addressed. The `meta` scope should probably be tested on a few of the different forms to make sure it terminates properly.

wbond commented on Scala/syntax_test_scala.scala:
```diff
@@ -1189,3 +1189,38 @@ def <(a: Int) = 42
+// ^^ storage.modifier.annotation
+//            ^^ string
+
+@transient @volatile var m: Int
```
Add assertions for the second annotation here to ensure we a properly allowing two back-to-back.

djspiewak commented:
@gwenzek Can you add a test for the following?

```scala
x: Foo @volatile with Bar
```

I think my comment from earlier got lost in the wash.  I'm not sure the above will correctly highlight with your current implementation.

gwenzek commented:
@wbond  I added your suggestions 

@djspiewak done and it works

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
       push: single-type-expression
 
+  annotation:
+    - match: '(@){{plainid}}\.'
```
Sorry, missed this the first time: the `.` should be `punctuation.accessor.scala`.

FichteFoll replied:
How about specializing this to `punctuation.accesor.dot.scala`, like it's done for comment punctuation?

I suggested this a few months ago already for C# iirc.

gwenzek replied:
done.

@FichteFoll I kept "punctuation.accessor.scala" in order to stay coherent with the rest of the file. This scope renaming could go in a separate PR.

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+        1: punctuation.definition.annotation.scala
+      push:
+        - meta_scope: meta.annotation.scala
+        - match: '{{plainid}}\.'
```
Again, `punctuation.accessor.scala` on the `.`.

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+  annotation:
+    - match: '(@){{plainid}}\.'
+      captures:
+        0: storage.modifier.annotation.scala
```
Stylistically we've been moving away from `0` captures and instead using `scope: ` for when a scope applies to the whole match.

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+          set: annotation-parameters
+    - match: '(@){{plainid}}'
+      captures:
+        0: storage.modifier.annotation.scala
```
Move to `scope: `.

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+  annotation-parameters:
+    - meta_scope: meta.annotation.scala
+    - match: \(
+      scope: punctuation.definition.arguments.annotation.begin.scala
```
Need this to be `punctuation.section.arguments...`

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+        - match: ','
+          scope: punctuation.separator.arguments.annotation.scala
+        - match: \)
+          scope: punctuation.definition.arguments.annotation.end.scala
```
Also `punctuation.section.arguments...`.

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+          pop: true
+        - include: main
+    - match: \[
+      scope: punctuation.definition.arguments.annotation.begin.scala
```
Another `punctuation.section...`

gwenzek replied:
done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+      scope: punctuation.definition.arguments.annotation.begin.scala
+      push:
+        - match: \]
+          scope: punctuation.definition.arguments.annotation.end.scala
```
`punctuation.section...`

gwenzek replied:
Done

wbond commented on Scala/Scala.sublime-syntax:
```diff
@@ -147,6 +147,46 @@ contexts:
+          scope: punctuation.definition.arguments.annotation.end.scala
+          pop: true
+        - include: delimited-type-expression
+    - match: \s*
```
I would probably move this as part of the next negative lookahead, e.g. `(?![(\[\s])`

gwenzek commented:
@wbond sorry I shouldn't have committed at 1am ! I'll finish that soon.

gwenzek commented:
Hi @wbond , I think this version should be ok.
I added tests for the "meta.scope" tell me if you think it's enough.

djspiewak commented:
The following is incorrectly highlighted with this PR:

```scala
x: Foo @abc with Bar
```

To see why it is incorrect, remove the annotation and note the scoping on Bar.

djspiewak commented:
@gwenzek I've PR'd (to you) a fix for the test case I referenced.

@wbond I've been dogfooding this branch for a few months now.  Aside from the aforementioned broken case (and its fix), it really looks ready to go.

gwenzek commented:
Thanks Daniel, I'll have a look a it soon.

I'm sorry I didn't took the time to finish that, I've had less time for
this kind of thing this last months.

wbond commented:
Sorry for the delay on this, and thanks for your work on it.

We did some discussion about annotation scopes at #709. I made a decision after I gathered some feedback: https://github.com/sublimehq/Packages/issues/709#issuecomment-266835130.

If you want to update this PR with those, great! If you won't have time for a while, I can pull this in and make the scope name tweaks myself.

djspiewak commented:
@wbond If you do end up pulling this in manually, be sure to grab my fix! (mentioned above) Without it, certain types involving `with` will be incorrectly scoped.

wbond commented:
@djspiewak Will do, thanks!

gwenzek commented:
I included @djspiewak fix, renamed the annotation scope to `variable.annotation` and rebased on master.
I didn't introduce `meta.annotation.identifier` nor `meta.annotation.parameters` but  the first one would be equivalent to `meta.annotation variable.annotation` .

wbond commented:
@gwenzek Thanks for the tweaks!

The `meta.annotation.identifier` scope primarily exists for a multi-part identifier that includes accessors. I suppose it depends on the language for if it is needed. However, it probably wouldn't be a bad idea to use `meta.annotation.identifier variable.annotation` for a single identifier token. This way plugins can rely on always using `meta.annotation.identifier` to fully resolve the annotation name.

For instance, with Python a decorator could be `@module.ClassName.static_method(True)`. The entirety of `module.ClassName.static_method` would get `meta.annotation.identifier`, while only `static_method` would get `variable.annotation`, and the `.` would get `punctuation.accessor`. The part `(True)` would get `meta.annotation.parameters`, `(` would get `punctuation.section.parens.begin` and `True` would get `constant.language`.

Does that make sense?

FichteFoll commented:
I agree, except that `(` should get `punctuation.section.arguments.begin`.

wbond commented:
> I agree, except that ( should get punctuation.section.arguments.begin.

Currently the guidelines suggest `group` or `parens`. For most of the paired punctuation we've stayed away from semantic names since the semantics of the symbols can vary greatly between syntaxes. I guess the exception so far has been `generic`. I think in some situations we've used `angle` or `angles`.

FichteFoll commented:
Hm, I took that from the Python syntax definition after you worked on it. It uses `punctuation.section.parameters` and (in a PR of mine) `punctuation.section.arguments`, similar to how `meta.arguments` is used. I don't really care what we chose to do since all these punctiation tokens most likely also have a meta-scope associated with them, though.

wbond commented:
Yes, I am sure there are syntaxes that have been worked on in the past year that break the guidelines. I know Rust needs to be revisited for some things, and I imagine there are probably some tweaks that should be made in C/C++ and JavaScript.

> This reads more like an argument for my position, honestly.

As in, let's not use names for what they mean in the language, but rather what symbol they are. That way color schemes don't need to handle `arguments`, `parameters`, `tuple`, `group`, `cast`, etc. Hopefully they can just target `parens` and `group` and be done.

I suppose some color scheme authors will want to highlight `(`, `[` `{` all the same for macro parameters in Rust (they are all viable), which would be an argument for `parameters`. But I guess is someone really wants to do that they should be able to use a meta scope to target specific instances of `punctuation.section`.

----

Overall I think the best we can do is try to be consistent, and I think more abstract names are probably better, but this isn't a position I hold too strongly.

FichteFoll commented:
So, this is kind of related to what we discussed in #727, except it appears that I'm of the opposite opinion in this one. This actually makes me kind of wonder whether the decision in #727 of using `punctuation.accessor.dot` is the better one after all ... I'll need to think more about this.

I don't quite get `group`, though, which *is* semantic.

By the way, the guidelines don't have anything for `definition` vs `scope` either and you probably want to add that too when working on them.

wbond commented:
@FichteFoll Let's open a new issue if you want to continue discussing more ideas for clarifying the guidelines.

For now I am going to merge this and we can tweak it later as need be.

wbond commented:
I made some tweaks in eaf8321 that apply `meta.annotation.identifier` as I had originally intended, removed `variable.annotation` from the `@` symbol and only applies `variable.annotation` to the final identifier in the annotation name.

wbond commented:
@gwenzek Thanks for all of your work on this!
